planner/core/exhaust_physical_plans.go

```diff
@@ -2493,15 +2493,40 @@ func (p *LogicalLock) exhaustPhysicalPlans(prop *property.PhysicalProperty) ([]P
 
 func (p *LogicalUnionAll) exhaustPhysicalPlans(prop *property.PhysicalProperty) ([]PhysicalPlan, bool) {
 	// TODO: UnionAll can not pass any order, but we can change it to sort merge to keep order.
-	if !prop.IsEmpty() || prop.IsFlashProp() {
+	if !prop.IsEmpty() || (prop.IsFlashProp() && prop.TaskTp != property.MppTaskType) {
+		return nil, true
+	}
+	if prop.TaskTp == property.MppTaskType && prop.PartitionTp != property.AnyType {
 		return nil, true
 	}
+	canUseMpp := p.ctx.GetSessionVars().AllowMPPExecution && p.canPushToCop(kv.TiFlash)
```

Comment:
this function does not deliver the partition properity, which introduce more network overhead during exchange.
It is better to deliver the partition properity, then the unionAll becomes a exchange reciver, and set different exchange sender such as pashthrough or partition. for example:
group by R.a
|
(group by S.a, S.b union scan C) as R

if use the current code, the plan becomes:

group by R.a
|
exchange receiver
|----exchange sender1[partition by S.a] - group by S.a, S.b[partition by S.a,S.B] 
|----exchange sender2[partition by C.a] - Scan C

the optimal one is:

group by R.a
|
exchange receiver
|----exchange sender1[pashthrough] - group by S.a, S.b[partition by S.a] 
|----exchange sender2[partition by C.a] - Scan C

here exchange sender1 will send to local node and reduce network IO.

Reply:
That's too complicated and only is beneficial in a few situations.

Reply:
eliminating shuffle is an important optimization for distributed plans. If the users' sql is slow due to more unnecessary exchange, how to support this case in the future?
 besides, it is not too complicated, its main idea is:

0. delivery `data distribution properity` to its child,
1. push up unionAll until the data distribution properity does not satisfy needs, then
2. enforce  exchange.

Reply:
Chould you detail the pass through exchanger? How could we granrantee the same task number between source tasks and target tasks?